### PR TITLE
add gather_nd op and unit test

### PR DESCRIPTION
**Pull request by ForFishes**
Add gather_nd op implementation and the unit test codes of it. It is actually a high-dimensional extension of gather op and supports for simultaneous indexing by multiple axes.

### CODE REVIEW

**Comment by ForFishes** [comment]
image
image

**Review by hutuxian — DISMISSED** [review]

**Review by hutuxian — APPROVED** [review]
LGTM for API.

File: python/paddle/fluid/tests/unittests/test_gather_nd_op.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    Index has empty element, which means copy entire tensor
+    """
+
+    def setUp(self):
+        self.op_type = "gather_nd"
+        xnp = np.array(
+            [[65, 17, 2], [-14, -25, -1], [76, 22, 3]]).astype("float32")
+        self.inputs = {'X': xnp, 'Index': np.array([[], []]).astype("int32")}
+        self.outputs = {
+            'Out': np.vstack((xnp[np.newaxis, :], xnp[np.newaxis, :]))
```
Comment (gavin1332) [inline python/paddle/fluid/tests/unittests/test_gather_nd_op.py:34]: why use newaxis?

File: python/paddle/fluid/tests/unittests/test_gather_nd_op.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    """
+
+    def setUp(self):
+        self.op_type = "gather_nd"
+        xnp = np.array(
+            [[65, 17, 2], [14, 25, 1], [76, 22, 3]]).astype("float64")
+        index = np.array([[1, 1], [2, 1]]).astype("int64")
+
+        self.inputs = {'X': xnp, 'Index': index}
+        self.outputs = {'Out': xnp[tuple(index.T)]}  #[25, 22]
```
Comment (gavin1332) [inline python/paddle/fluid/tests/unittests/test_gather_nd_op.py]: should be [25, 76] not [25, 22]?

File: python/paddle/fluid/tests/unittests/test_gather_nd_op.py (hunk 1 above)
Comment (gavin1332) [reply python/paddle/fluid/tests/unittests/test_gather_nd_op.py:34]: to add an extra dimension

**Review by gavin1332 — APPROVED** [review]
LGTM

**Review by frankwhzhang — APPROVED** [review]

**Pull request merged**